File: day14/day14.py
# Import modules
import os
from termcolor import colored
from collections import Counter
import threading

# Input parser
def input_parser(filename):
    # Open the file
    __location__ = os.path.realpath(os.path.join(os.getcwd(), os.path.dirname(__file__)))
    inputFile = open(os.path.join(__location__, filename), 'r')

    # Process the file
    input = inputFile.readlines()
    template = list(input[0][:-1])
    rules = [line[:-1].split(' -> ') for line in input[2:]]

    # Close the file and return the result
    inputFile.close()
    return (template, rules)

# Puzzle 1
def puzzle1(filename):
    # Read file
    (polymer, rules) = input_parser(filename)

    # Apply the rules 10 times
    for _ in range(10):
        oldPolymer = polymer.copy()
        polymerIdx = 0
        for oldPlymerIdx in range(len(oldPolymer)-1):
            twoChars = ''.join(oldPolymer[oldPlymerIdx:oldPlymerIdx+2])
            if twoChars in [rule[0] for rule in rules]:
                ruleIdx = [rule[0] for rule in rules].index(twoChars)
                polymer.insert(polymerIdx+1, rules[ruleIdx][1])
                polymerIdx += 2
            else:
                polymerIdx += 1

    # Count the number of occurances
    numOccurances = Counter(polymer).most_common()
    return numOccurances[0][1] - numOccurances[-1][1]

# Puzzle 2
def applyRules(polymer, rules):
    oldPolymer = polymer.copy()
    polymerIdx = 0
    for oldPlymerIdx in range(len(oldPolymer)-1):
        twoChars = ''.join(oldPolymer[oldPlymerIdx:oldPlymerIdx+2])
        if twoChars in [rule[0] for rule in rules]:
            ruleIdx = [rule[0] for rule in rules].index(twoChars)
            polymer.insert(polymerIdx+1, rules[ruleIdx][1])
            polymerIdx += 2
        else:
            polymerIdx += 1

# Puzzle 2 solver
def puzzle2(filename):
    # Read file
    (polymer, rules) = input_parser(filename)

    # Apply the rules 40 times
    for i in range(40):
        # Split the polymer into up to 12 parts for up to 12 threads
        if len(polymer) < 12:
            numThreads = len(polymer)
        else:
            numThreads = 12
        chunkSize = int(len(polymer)/numThreads)
        polymerChunks = [polymer[j:j+chunkSize].copy() for j in range(0, len(polymer), chunkSize)]

        # Process the chunks
        threads = []
        for j in range(len(polymerChunks)):
            threads.append(threading.Thread(target=applyRules, args=(polymerChunks[j], rules)))
            threads[j].start()

        # Wait for the threads to finnish and combine them
        threads[0].join()
        polymer = polymerChunks[0].copy()
        for j in range(1,len(polymerChunks)):
            threads[j].join()
            twoChars = polymer[-1] + polymerChunks[j][0]
            if twoChars in [rule[0] for rule in rules]:
                ruleIdx = [rule[0] for rule in rules].index(twoChars)
                polymer.append(rules[ruleIdx][1])
            polymer.extend(polymerChunks[j])


        print(str(i))

    # Count the number of occurances
    numOccurances = Counter(polymer).most_common()
    return numOccurances[0][1] - numOccurances[-1][1]

# Run tests for puzzle 1
puzzle1TestPass = puzzle1('example1') == 1588
if(puzzle1TestPass):
    print(colored('Tests for puzzle 1 PASS', 'green'))
else:
    print(colored('Tests for puzzle 1 FAIL', 'red'))

# Solve puzzle 1 if test passed
if(puzzle1TestPass):
    print('Solution for puzzle 1: ' + str(puzzle1('input')))

# Run tests for puzzle 2
puzzle2TestPass = puzzle2('example1') == 1588
if(puzzle2TestPass):
    print(colored('Tests for puzzle 2 PASS', 'green'))
else:
    print(colored('Tests for puzzle 2 FAIL', 'red'))

# Solve puzzle 2 if test passed
if(puzzle2TestPass):
    print('Solution for puzzle 2: ' + str(puzzle2('input')))

File: day14/day14.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <array>
#include <map>

// Linked list version of a polynmer
class PolymerElement {
    public:
        char            element;
        PolymerElement  *nextElement;

        PolymerElement(char element, PolymerElement *nextElement) {
            this->element       = element;
            this->nextElement   = nextElement;
        }
};

class Polymer {
    public:
        PolymerElement  *head;
        PolymerElement  *tail;
        uint64_t        length;

        Polymer(PolymerElement *head, PolymerElement *tail, int length) {
            this->head      = head;
            this->tail      = tail;
            this->length    = length;
        }

        void print() {
            PolymerElement *currElement = head;
            while(currElement != NULL) {
                std::cout << currElement->element;
                currElement = currElement->nextElement;
            }
            std::cout << "\n";
        }

        void addElement(PolymerElement *currElement, PolymerElement *newElement)
        {
            PolymerElement *nextElement = currElement->nextElement;
            currElement->nextElement    = newElement;
            newElement->nextElement     = nextElement;
            this->length++;
            if(currElement == this->tail)
            {
                this->tail = newElement;
            }
        }
};

// Input parser
Polymer* inputParser(std::string filename, std::vector<std::array<char, 3>> &rules) {
    // Read the file
    std::vector<std::string> lines;
    std::fstream file;
    file.open(filename, std::ios::in);
    if(file.is_open()) {
        std::string line;
        while(getline(file, line)){
            lines.push_back(line);
        }
        file.close();
    }

    // Create the polymer from the template
    PolymerElement *head = new PolymerElement(lines[0][0], NULL);
    PolymerElement *currElement = head;
    int length = 1;
    for(int i = 1; i < lines[0].length(); i++) {
        PolymerElement *element = new PolymerElement(lines[0][i], NULL);
        currElement->nextElement = element;
        currElement = element;
        length++;
    }
    Polymer *polymer = new Polymer(head, currElement, length);

    // Fill in the list of rules
    for(int i = 2; i < lines.size(); i++) {
        std::array<char, 3> rule;
        rule[0] = lines[i][0];
        rule[1] = lines[i][1];
        rule[2] = lines[i][6];
        rules.push_back(rule);
    }

    return polymer;
}

int main() {
    // Read the file
    std::vector<std::array<char, 3>> rules;
    Polymer *polymer = inputParser("example1", rules);
   
    // Apply the rule 40 times
    for(int i = 0; i < 40; i++) {
        PolymerElement *currElement = polymer->head;
        while(currElement != polymer->tail) {
            for(int j = 0; j < rules.size(); j++) {
                if(currElement->element == rules[j][0] && currElement->nextElement->element == rules[j][1]) {
                    polymer->addElement(currElement, new PolymerElement(rules[j][2], NULL));
                    currElement = currElement->nextElement;
                    break;
                }
            }
            currElement = currElement->nextElement;
        }

        std::cout << i;
        std::cout << "\n";
    }

    // Count the number of each element
    std::map<char, uint64_t> numOccuurances;
    PolymerElement *currElement = polymer->head;
    while(currElement != NULL) {
        if(numOccuurances.find(currElement->element) == numOccuurances.end()) {
            numOccuurances.insert(std::make_pair(currElement->element, 1));
        }
        else {
            uint64_t prevOcc = numOccuurances[currElement->element];
            numOccuurances[currElement->element] = prevOcc + 1;
        }
        currElement = currElement->nextElement;
    }

    // Find the number of occurances that is most and least common
    uint64_t mostCommonOcc = 0;
    uint64_t leastCommonOcc = -1;
    std::map<char, uint64_t>::iterator mapIterator;
    for(mapIterator = numOccuurances.begin(); mapIterator != numOccuurances.end(); mapIterator++) {
        if(mapIterator->second > mostCommonOcc) {
            mostCommonOcc = mapIterator->second;
        }
        if(mapIterator->second < leastCommonOcc) {
            leastCommonOcc = mapIterator->second;
        }
    }

    // Print result
    std::cout << mostCommonOcc - leastCommonOcc;

    return 0;
}